### Imports

In [1]:
import pandas as pd

### Setup

In [2]:
# === Constantes
FILE_NAME = 'historico.xlsx'

### Caregando dados

In [3]:
# === Dataframe
df = pd.read_excel(FILE_NAME)

# === Sub dataframes importantes
# Alunos reprovados (Média < 5.0)
df_reprovados = df[df['Média'] < 5.0]

# === Valores importantes
# Total de alunos no dataset
total_alunos = len(df['Matrícula'].unique())

---

### Ex 1 

Qual é, percentualmente, a taxa de alunos que reprovam em uma disciplina pela 1º vez? E a mesma disciplina pela 2ª vez? 3ª? 4ª? 5ª?...

In [131]:
# Contabilizando o total de reprovacoes por disciplina, por aluno
df_contagem_reprovacoes = df_reprovados.groupby(['Matrícula', 'Disciplina'])['Matrícula'].count()
for i in range(1,6):
    # Separa as ocorrencias de i ou mais reprovaçoes na mesma materia por um dado aluno
    reprovados_i_vezes = df_contagem_reprovacoes[df_contagem_reprovacoes >= i]
    # Extrai a matriculas dos alunos, eliminando duplicados
    matriculas = reprovados_i_vezes.index.get_level_values('Matrícula').unique()
    # Conta o numero de matriculas
    numero_de_alunos = len(matriculas)
    # Exibindo o resultado
    print(f'{numero_de_alunos:>3}/{total_alunos} ({numero_de_alunos/total_alunos*100:>2.0f}%) dos alunos ja reprovaram ao menos {i} vez(es) em uma mesma disciplina')



827/1201 (69%) dos alunos ja reprovaram ao menos 1 vez(es) em uma mesma disciplina
441/1201 (37%) dos alunos ja reprovaram ao menos 2 vez(es) em uma mesma disciplina
239/1201 (20%) dos alunos ja reprovaram ao menos 3 vez(es) em uma mesma disciplina
107/1201 ( 9%) dos alunos ja reprovaram ao menos 4 vez(es) em uma mesma disciplina
 50/1201 ( 4%) dos alunos ja reprovaram ao menos 5 vez(es) em uma mesma disciplina


---

### Ex 2

Qual disciplina reprova mais? Qual reprova menos?

In [180]:
indice_reprovacao = df_reprovados['Disciplina'].value_counts(normalize=True, sort=True, ascending=False)*100
maior, menor = indice_reprovacao.iloc[[0,-1]].items()

print("Disciplina | Indice de reprovação (%)")
print(f"{maior[0]:^10d} | {maior[1]:^22.4f}")
print(f"{menor[0]:^10d} | {menor[1]:^22.4f}")

# indice_significante = indice_reprovacao[indice_reprovacao > 1]
# indice_significante.loc['Outros'] = 100-indice_significante.sum()
# indice_significante.plot(kind='pie', figsize=(10,10));

Disciplina | Indice de reprovação (%)
  41082    |        9.12964        
  56481    |        0.01522        


O indice de reprovação calculado acima indica, do total reprovações, quantos % são relativos a determinada disciplina.

- A disciplina responsável pelo maior número de reprovações (9.1296%) é a de código 41082.
- A disciplina responsável pelo menor número de reprovações (0.0152%) é a de código 56481.

---

### Ex 3

Qual é a relação entre a nota obtida na reprovação e o sucesso/fracasso da próxima vez que o aluno faz a mesma disciplina?

In [112]:
df= pd.DataFrame([["potatoes",20],["carots",39], ["tomatos",40], ["apples",2], ["bananas",2]] , columns = ["aliments","number"])  

df_draw = df.copy()
df_draw.loc[df_draw['number'] < 20, 'aliments'] = 'vegetables'

df_draw = df_draw.groupby('aliments')['number'].sum().reset_index()

df

aliments  number
0  potatoes      20
1    carots      39
2   tomatos      40
3    apples       2
4   bananas       2